homeassistant/components/mqtt/__init__.py

```diff
@@ -800,16 +813,69 @@ def _mqtt_on_disconnect(self, _mqttc, _userdata, result_code: int) -> None:
             _LOGGER.warning(
                 "Disconnected from MQTT (%s). Trying to reconnect in %s s",
                 result_code, wait_time)
-            # It is ok to sleep here as we are in the MQTT thread.
-            time.sleep(wait_time)
+
+            try:
+                await asyncio.sleep(wait_time)
+            except asyncio.CancelledError:
+                break
             tries += 1
 
+    def _mqtt_on_socket_open(self, client, userdata, sock):
+        """Socket open callback."""
+        self.hass.add_job(self._handle_socket_open, sock)
```

Comment:
We don't need to add a job. Just follow the example.

Reply:
We do, add this might be called by the `connect` or `reconnect` worker thread

Reply:
Well, it's only needed because `connect` and `reconnect` are called from worker threads because they block so the example can't be blamed I guess..
Comments added.

Reply:
But blocking I/O should never be done in the event loop. So that part of the example seems bad. And with all the threading locks in the code, I'm not sure how they think an asyncio loop should be used as external loop.

Reply:
Yeah.. So do you suggest to just drop this PR?

Reply:
The one that is most important is that the message handler is async. The connection open/close are only happening once